PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/appservice/SSHTerminalManager.java

```diff
@@ -0,0 +1,203 @@
+            DefaultLoader.getUIHelper().showError(SSH_INTO_WEB_APP_ERROR_MESSAGE, SSH_INTO_WEB_APP_ERROR_DIALOG_TITLE);
+            return connectionInfo;
+        }
+        connectionInfo.setOutputMessage(commendExecOutput.getOutputMessage());
+        connectionInfo.setSuccess(commendExecOutput.isSuccess());
+        if (commendExecOutput.isSuccess()) {
+            String username = PatternUtils.parseWordByPatternAndPrefix(commendExecOutput.getOutputMessage(), PatternUtils.PATTERN_WHOLE_WORD, "username: ");
+            if (StringUtils.isNotBlank(username)) {
+                connectionInfo.setUsername(username);
+            }
+            String port = PatternUtils.parseWordByPatternAndPrefix(commendExecOutput.getOutputMessage(), PatternUtils.PATTERN_WHOLE_NUMBER_PORT, "port: ");
+            if (StringUtils.isNotBlank(port)) {
+                connectionInfo.setPort(port);
+            }
+            String password = PatternUtils.parseWordByPatternAndPrefix(commendExecOutput.getOutputMessage(), PatternUtils.PATTERN_WHOLE_WORD, "password: ");
+            if (StringUtils.isNotBlank(password)) {
+                connectionInfo.setPassword(password);
+            }
+        }
+        return connectionInfo;
+    }
+
+    /**
+     * validate create-remote-connection output to ensure it's ready to ssh to local proxy and open terminal.
+     *
+     * @param connectionInfo
+     * @param webAppName
+     */
+    public boolean validateConnectionOutputBeforeOpenInTerminal(CreateRemoteConnectionOutput connectionInfo, String webAppName) {
+        if (connectionInfo == null) {
+            DefaultLoader.getUIHelper().showError(String.format(SSH_INTO_WEB_APP_ERROR_MESSAGE_WITH_NAME, webAppName), SSH_INTO_WEB_APP_ERROR_DIALOG_TITLE);
+            return false;
+        }
+        if (connectionInfo.isSuccess()) {
+            if (StringUtils.isBlank(connectionInfo.getPort())) {
+                DefaultLoader.getUIHelper().showError(String.format(SSH_INTO_WEB_APP_ERROR_MESSAGE_WITH_NAME, webAppName), SSH_INTO_WEB_APP_ERROR_DIALOG_TITLE);
+                return false;
+            }
+        } else {
+            DefaultLoader.getUIHelper().showError(String.format(SSH_INTO_WEB_APP_ERROR_MESSAGE_WITH_NAME, webAppName), SSH_INTO_WEB_APP_ERROR_DIALOG_TITLE);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * ssh to connect to local proxy and open the terminal for remote container.
+     *
+     * @param shellTerminalWidget
+     * @param connectionInfo
+     */
+    public void openConnectionInTerminal(ShellTerminalWidget shellTerminalWidget, CreateRemoteConnectionOutput connectionInfo) {
+        try {
+            int count = 0;
+            while ((shellTerminalWidget.getTtyConnector() == null) && count++ < 200) {
+                Thread.sleep(100);
+            }
+            shellTerminalWidget.executeCommand(String.format(CMD_SSH_TO_LOCAL_PROXY, connectionInfo.getPort()));
+            Thread.sleep(5000);
```

Comment:
why sleep 5 seconds here? can we replace it with test and retry logic?

Reply:
we can't. it's only three chances for password retry during ssh connecting.